clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -3302,14 +3319,24 @@ void Sema::AddSYCLIntelNumSimdWorkItemsAttr(Decl *D,
         return;
       }
     }
-
-    // If the declaration has an [[intel::reqd_work_group_size]] attribute,
-    // check to see if the first argument can be evenly divided by the
-    // num_simd_work_items attribute.
+    // If the declaration has a [[intel::reqd_work_group_size()]]
+    // attribute, check to see if the last argument can be evenly
+    // divided by the num_simd_work_items attribute.
+    // If the declaration has a __attribute__((reqd_work_group_size))
+    // or [[cl::reqd_work_group_size()]] attribute, check to see if the
+    // first argument can be evenly divided by the num_simd_work_items
+    // attribute. GNU and [[cl::reqd_work_group_size()]] spelling of
+    // ReqdWorkGroupSizeAttr maps to the OpenCL semantics. First and last
+    // argument are only swapped for the Intel atributes in SYCL and not
+    // the OpenCL ones.
     if (const auto *DeclAttr = D->getAttr<ReqdWorkGroupSizeAttr>()) {
       Optional<llvm::APSInt> XDimVal = DeclAttr->getXDimVal(Context);
+      Optional<llvm::APSInt> ZDimVal = DeclAttr->getZDimVal(Context);
 
-      if (*XDimVal % ArgVal != 0) {
+      llvm::APSInt WorkGroupSize =
```

Comment:
We might have a testing hole here with a bug hiding in it -- won't this wind up with `ZDimVal` being `llvm::None` because the attribute has not yet been instantiated?
```
template <int X, int Y, int Z>
[[intel::reqd_work_group_size(X, Y, Z)]]
void func();

template <int X, int Y, int Z>
[[intel::num_simd_work_items(3)]] void func(); // When we go to process this attribute.
```